proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,473 @@
+``writeMutVar#``, and a GC write barrier (the equivalent of
+``dirty_MUT_VAR()``).
+
+
+Transformations
+^^^^^^^^^^^^^^^
+
+Because the constructor worker for a mutable constructor is a stateful
+operation, GHC can no longer assume that an expression like ``$wK e1...en``
+has type ``K t1...tn`` when ``K`` is a mutable constructor.  This
+assumption is currently used in a couple of places:
+
+- In Worker-wrapper, we build an expression representing the re-packed
+  constructor.  Worker-wrapper would need to be either disabled
+  (easiest) or adapted for mutable constructors.
+- When simplifying a case expression like ``case x of y { C a b -> E
+  }``, GHC creates the mapping ``y -> C a b`` when simplifying ``E``.
+  We will have to avoid creating this mapping If ``C`` is a mutable
+  constructor.
+
+
+
+Runtime System
+~~~~~~~~~~~~~~
+
+Garbage collection
+^^^^^^^^^^^^^^^^^^
+
+The garbage collector needs to know that an object is mutable, and
+which fields are mutable.  So we have to put this information in the
+info table.  Probably:
+
+- new ``MUT_CONSTR_CLEAN`` and ``MUT_CONSTR_DIRTY`` object types
+
+- Each constructor with a mutable field needs two info tables: the
+  CLEAN one and the DIRTY one.  From each info table we need to be
+  able to get both CLEAN and DIRTY info pointers, for the GC write
+  barrier.
+
+- Representation: put the mutable fields first, before the non-mutable
+  pointers.
+
+- Store the number of mutable fields in the info table, alongside the
+  number of pointers and non-pointers.  (but include the mutable
+  fields in the count of pointer fields, so that RTS code that doesn't
+  care about mutability can work without changes)
+
+The GC needs to do the same CLEAN/DIRTY and non-eager promotion stuff
+that it does with other mutable objects.
+
+TODO: pin down the details of info table representation and the form
+of the GC write barrier code.
+
+Drawbacks
+---------
+
+The GC write barrier for a mutable constructor may be a little less
+efficient than the write barrier for a ``MutVar#``, but this is more
+than compensated for by losing a layer of indirection.
+
+Adding new syntax has lots of costs: changes in the parser and
+``HsSyn``, the renamer and typechecker, not to mention knock-on effects
+on external packages: ``haskell-src-exts`` and clients of that.
+
+Types that contain mutable fields cannot be UNPACKed into other
+constructors.
+
+Worker-wrapper doesn't work on mutable constructors, at least not
+without some changes.  Perhaps this isn't so bad, since we would never
+be able to eliminate the original construction of the mutable
+constructor anyway.
+
+Alternatives
+------------
+
+Don't do this :)
+
+Unresolved Questions
+--------------------
+
+* Can we add a way to include mutable arrays in a constructor?
+* It would be great to allow STM as an option in addition to IO and
+  ST.  The constructor will need to store extra metadata, because
+  TVar# is more complex than MutVar#.
```

Comment:
Sounds like another argument for PrimMonadLight. I'm having a hard time imagining what the intended semantics would look like for mutable fields in STM though...

Reply:
STM with mutable fields was the central part of my thesis and why I made an implementation.  The difficulty with pushing that to a high level with something like `PrimMonad` is we need extra fields in `STM` heap objects (at least that is a desirable design choice).